### PR TITLE
fix: address remaining review feedback

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary
- constrain GitHub/Okta live source base URLs, add HTTP timeouts/body limits, and persist source runtime progress after each synced page
- cap JetStream replay scans and tighten subject wildcard matching
- fix typed-nil claim timestamp handling, retract projected relation links, and map Connect errors to specific codes
- strengthen structural linters for back-pointer aliases/containers and remove sealed-interface dead code

## Validation
- go test ./sources/github ./sources/okta ./internal/appendlog/jetstream ./internal/sourceruntime
- go test ./internal/claims ./internal/statestore/postgres ./internal/bootstrap ./internal/graphstore/kuzu
- CGO_ENABLED=0 go test ./internal/graphstore/kuzu ./internal/bootstrap
- go -C tools/linters test ./nobackpointer ./sealedinterface
- make verify

### CODE REVIEW

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the GitHub Enterprise base_url, loopback SSRF, and Connect error-code findings are resolved, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the remaining Connect RPC error mapping and zoned IPv6 loopback findings are resolved, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fix. Focus on whether the bare query-delimiter base_url findings for GitHub and Okta are resolved, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the numeric IPv4 loopback and explicit retracted relation cleanup findings are resolved, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the same-batch relation reassertion and Okta tampering legacy finding ID findings are resolved, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the typed-nil reportStore/eventReplayer boundary and graph rebuild empty-page cursor findings are resolved, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed Connect error sanitization fix. Focus on whether internal Connect errors no longer expose underlying messages, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest timestamp fix on top of the updated branch. Focus on whether running finding evaluation runs now keep finished_at NULL and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest graph rebuild stage timing fix. Focus on whether dry-run stage confirmations no longer double-count read/replay time in project_graph, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest final edge-case fixes. Focus on whether projected-link deletion matches entity_links primary-key semantics, finding rule-selection/unsupported errors map to stable Connect codes, and the claims PK migration is safe under concurrent startups; only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the Kuzu projected-link delete identity fix. Focus on whether Kuzu link merge/delete semantics are now aligned with endpoint+relation identity, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the report HTTP error sanitization fix. Focus on whether writeReportError now returns 500/generic text for unexpected errors while preserving 400/404/503 for sentinels, and only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest validation-error mapping and typed-nil health dependency fixes. Focus on whether source/source-runtime/claim validation errors map to InvalidArgument, empty graph-ingest run IDs map to InvalidArgument, and componentStatus treats typed-nil pingers as unconfigured; only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest graph HTTP error and checkpoint hash hardening fixes. Focus on whether graph HTTP handlers no longer leak internal 5xx messages and graph-ingest checkpoint hashing ignores camelCase/common credential keys; only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest follow-up fixes for remaining HTTP error sanitization, Kuzu projected attribute merge/NULL handling, and claims primary-key migration locking. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid please re-review the latest follow-up hardening commit. It preserves invalid-request mapping for report source-config errors and serializes Kuzu projected-link deletes with projection updates.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after fixing the latest findings: client parse/body/query validation errors are classified as 400 without leaking internals, and Kuzu projected links to missing endpoints are ignored/no-op instead of failing projection. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the tenant-change follow-up: SourceRuntime Put now rejects tenant_id changes for existing runtime IDs to avoid stale cross-tenant claims. The previous parse-error and missing-endpoint fixes remain in place. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid please re-review the latest fixes for HTTP parse-error classification, Kuzu missing-endpoint links, and runtime tenant-change rejection.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after blocking redirect-based SSRF in the GitHub and Okta source HTTP clients. The clients now use no-redirect policies and tests assert redirect targets are not followed. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after moving Jira workspace_key validation before SDK ensure_runtime/ensureRuntime side effects in both Python and TypeScript. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid please re-review after adding findings.ErrInvalidRequest and mapping finding validation failures to InvalidArgument/400. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid please re-review after preserving source/source-runtime validation errors as InvalidRequest and tightening Okta domain host-only validation. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid please re-review after tightening the TypeScript Jira SDK identifier normalization to reject object/array coercion. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid please re-review after rejecting object-coerced Jira SDK identifiers and removing the unreachable GitHub audit user-scope branch. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid please re-review after rejecting non-scalar Python Jira identifiers and preserving /api/v3 for api.* GitHub Enterprise base URLs. Only report remaining high-confidence actionable issues.

**Comment by jonathanhaaswriter** [comment]
### Droid sweep status

Latest sweep: `2026-04-29 22:09 UTC`.

- Unresolved non-outdated review threads: `0`
- CI `verify`: `SUCCESS`
- `droid-review`: `SUCCESS`

I’ll keep watching for new review findings/check failures and address them as they appear.

**Comment by jonathanhaaswriter** [comment]
Droid sweep audit complete (final pass)

- Re-scanned all live review threads, including outdated ones
- Verified each remaining thread against the current branch tip
- Resolved every actionable thread (or fixed it inline and pushed)
- 0 unresolved review threads across the whole open-PR stack at this time

PR #494 (fix/feedback-main-followup) and PR #495 (feat/droid-creation-workflows) received additional fixes for SSRF (DNS-rebind) and workflow branch-name validation in this pass.

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid please harden this PR end-to-end:

1. Re-review the diff for residual bugs, races, edge cases, security gaps, and untested paths.
2. Apply the smallest correct fixes for any high-confidence issues you find.
3. Strengthen test coverage: add regression tests for every prior review finding on this PR plus the new behavior, including failure paths, boundary inputs, concurrency, and not-found / typed-nil interface guards.
4. Run focused tests and `make verify`. If validators fail, fix until green.
5. Commit and push the fixes to this PR's branch (or, if you lack push permission in this workflow, leave concrete inline suggestions and a summary comment).

Do not change the PR scope, do not modify the default branch, and do not merge.

**Comment by jonathanhaaswriter** [comment]
@droid review

File: internal/bootstrap/app.go [hunk 1]
```diff
@@ -321,7 +325,7 @@ func (a *App) handleRunReport(w http.ResponseWriter, r *http.Request) {
 	configReq.Header.Del("X-Cerebro-Source-Config")
 	config, err := sourceConfigFromRequest(configReq)
 	if err != nil {
-		writeReportError(w, err)
+		writeReportError(w, fmt.Errorf("%w: %w", reports.ErrInvalidRequest, err))
 		return
 	}
 	for key, value := range config {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/bootstrap/app.go:331]: Already fixed on this branch: report parameters are normalized in `reports.Service.Run()` and `normalizeParameters` drops sensitive keys such as token/secret/password before persistence; existing regression `TestNormalizeParametersDropsSensitiveKeys` covers this.

File: .github/workflows/droid.yml [hunk 2]
```diff
@@ -15,11 +15,22 @@ on:
...
+      (github.event_name == 'pull_request_review' &&
+        contains(fromJSON('["OWNER","MEMBER","COLLABORATOR"]'), github.event.review.author_association) &&
+        contains(github.event.review.body, '@droid')) ||
+      (github.event_name == 'issues' &&
+        contains(fromJSON('["OWNER","MEMBER","COLLABORATOR"]'), github.event.issue.author_association) &&
+        (contains(github.event.issue.body, '@droid') || contains(github.event.issue.title, '@droid'))) ||
+      (github.event_name == 'pull_request' &&
+        github.event.pull_request.head.repo.full_name == github.repository &&
+        contains(fromJSON('["OWNER","MEMBER","COLLABORATOR"]'), github.event.pull_request.author_association) &&
+        (contains(github.event.pull_request.body, '@droid') || contains(github.event.pull_request.title, '@droid')))
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply .github/workflows/droid.yml]: Already fixed on this branch: `.github/workflows/droid.yml` coalesces nullable webhook body/title fields with `|| ''` before `contains()`, including issue, PR, review, and review-comment paths.

File: sources/github/source.go [hunk 3]
```diff
@@ -323,6 +388,161 @@ func parseSettings(cfg sourcecdk.Config, requireRepo bool) (settings, error) {
...
+		}
+		return errUnsafeBaseURLHost
+	}
+	if lookupIPAddrs == nil {
+		lookupIPAddrs = net.DefaultResolver.LookupIPAddr
+	}
+	addrs, err := lookupIPAddrs(ctx, normalized)
+	if err != nil {
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sources/github/source.go:488]: Fixed in f8789bb8d by failing closed when DNS lookup fails in `rejectResolvedUnsafeHost`; added a regression ensuring the wrapped transport is not reached after lookup failure.

File: internal/sourceops/service.go [hunk 4]
```diff
@@ -42,7 +45,7 @@ func (s *Service) Check(ctx context.Context, req *cerebrov1.CheckSourceRequest)
 		return nil, err
 	}
 	if err := source.Check(ctx, sourcecdk.NewConfig(req.GetConfig())); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%w: %w", ErrInvalidRequest, err)
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceops/service.go]: Fixed in 9df1f4f0e: source operations now only map errors marked as invalid source config to InvalidArgument; upstream/live source failures pass through as retriable server errors. Added regression coverage and ran make check.

File: internal/appendlog/jetstream/jetstream.go [hunk 5]
```diff
@@ -170,15 +181,20 @@ func (l *Log) Replay(ctx context.Context, req ports.ReplayRequest) ([]*cerebrov1
 	if stream.State.LastSeq == 0 || stream.State.LastSeq < stream.State.FirstSeq {
 		return events, nil
 	}
+	scanned := 0
 	for seq := stream.State.FirstSeq; seq <= stream.State.LastSeq; seq++ {
+		if scanned >= maxReplayScan {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/appendlog/jetstream/jetstream.go]: Fixed in 9df1f4f0e: removed the hard replay scan cutoff so sparse matches beyond 10k stream messages can still be returned. Added sparse-stream replay coverage and ran make check.

File: internal/claims/service.go [hunk 6]
```diff
@@ -127,6 +132,12 @@ func (s *Service) WriteClaims(ctx context.Context, request WriteRequest) (*Write
 				result.EntitiesUpserted++
 			}
 		}
+		if retracted := retractedRelation(runtime, claim); retracted != nil {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/claims/service.go:133]: Fixed in 9df1f4f0e: relation claim normalization now rejects subject/object URNs outside the runtime tenant before any projected-link deletion can run. Added cross-tenant retraction coverage and ran make check.

File: .github/workflows/droid.yml [hunk 7]
```diff
@@ -15,11 +15,22 @@ on:
...
-      (github.event_name == 'pull_request' && (contains(github.event.pull_request.body, '@droid') || contains(github.event.pull_request.title, '@droid')))
+      (github.event_name == 'issue_comment' &&
+        contains(fromJSON('["OWNER","MEMBER","COLLABORATOR"]'), github.event.comment.author_association) &&
+        contains(github.event.comment.body || '', '@droid')) ||
+      (github.event_name == 'pull_request_review_comment' &&
+        contains(fromJSON('["OWNER","MEMBER","COLLABORATOR"]'), github.event.comment.author_association) &&
+        contains(github.event.comment.body || '', '@droid')) ||
+      (github.event_name == 'pull_request_review' &&
+        contains(fromJSON('["OWNER","MEMBER","COLLABORATOR"]'), github.event.review.author_association) &&
+        contains(github.event.review.body || '', '@droid')) ||
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply .github/workflows/droid.yml]: Fixed in 350d1288e: review-comment and review-triggered @droid runs now require the PR head repo to match the base repository, matching the pull_request fork guard. Validated with make check.

File: internal/sourceops/service.go [hunk 8]
```diff
@@ -111,7 +121,7 @@ func (s *Service) lookup(sourceID string) (sourcecdk.Source, error) {
 func previewEvents(events []*cerebrov1.EventEnvelope) ([]*cerebrov1.SourcePreviewEvent, error) {
 	previews := make([]*cerebrov1.SourcePreviewEvent, 0, len(events))
 	for _, event := range events {
-		preview := &cerebrov1.SourcePreviewEvent{Event: event, EventId: event.GetId()}
+		preview := &cerebrov1.SourcePreviewEvent{EventId: event.GetId()}
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceops/service.go]: Fixed in 350d1288e: SourcePreviewEvent now preserves the full event field while keeping event_id and decoded payload fields populated. Added/updated sourceops coverage and ran make check.

File: sdk/typescript/test/jira.test.mjs [hunk 9]
```diff
@@ -0,0 +1,56 @@
+import assert from "node:assert/strict";
+import { readFile } from "node:fs/promises";
+import { stripTypeScriptTypes } from "node:module";
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sdk/typescript/test/jira.test.mjs]: Fixed in 350d1288e: the TypeScript test no longer imports node:module stripTypeScriptTypes, so it runs under the supported Node 20 runtime. Verified with node --test sdk/typescript/test/jira.test.mjs and make check.

**Comment by jonathanhaaswriter** [comment]
@droid review

Please re-review the latest head deeply, not just the touched lines. Focus especially on:

1. **Workflow-trigger security**: verify every `@droid`-triggering path (`issue_comment`, `pull_request_review_comment`, `pull_request_review`, `pull_request`) cannot execute privileged review actions for forked PR code or attacker-controlled refs, and that `github.event.*` shapes are safe for each event type.
2. **Source error taxonomy**: confirm caller-fixable source config errors still become `InvalidArgument`, while transient upstream/network/DNS/provider failures remain retriable server errors across HTTP and Connect paths.
3. **SSRF/DNS rebinding**: re-check GitHub base URL validation at config time, redirect time, and dial time; include resolver failures, literal IPs, IPv6, rebinding, enterprise `/api/v3` paths, and custom transports.
4. **Replay correctness and resource safety**: after removing the hard scan cutoff, look for unbounded or sparse-stream replay risks, context cancellation handling, JetStream missing-message behavior, and whether callers have sufficient limits/cursors.
5. **Tenant isolation**: validate claim relation retractions and projected-link deletes cannot cross tenant/runtime boundaries, especially now that deletes are keyed by relation endpoints.
6. **API compatibility**: ensure `preview_events[*].event`, `event_id`, decoded payloads, and generated protos stay backward compatible for all consumers.
7. **SDK tests/runtime**: ensure TypeScript tests run on the repo-supported Node version and still catch object-coercion regressions without relying on unavailable experimental APIs.

File: .github/workflows/droid.yml [hunk 10]
```diff
@@ -15,11 +15,24 @@ on:
 jobs:
   droid:
     if: |
-      (github.event_name == 'issue_comment' && contains(github.event.comment.body, '@droid')) ||
-      (github.event_name == 'pull_request_review_comment' && contains(github.event.comment.body, '@droid')) ||
-      (github.event_name == 'pull_request_review' && contains(github.event.review.body, '@droid')) ||
-      (github.event_name == 'issues' && (contains(github.event.issue.body, '@droid') || contains(github.event.issue.title, '@droid'))) ||
-      (github.event_name == 'pull_request' && (contains(github.event.pull_request.body, '@droid') || contains(github.event.pull_request.title, '@droid')))
+      (github.event_name == 'issue_comment' &&
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply .github/workflows/droid.yml]: Fixed in c1ef9389a by blocking issue_comment-triggered Droid runs on pull request conversations, leaving PR review/review-comment paths guarded by same-repo head checks.

File: internal/bootstrap/app.go [hunk 11]
```diff
@@ -1701,52 +1705,201 @@ func sensitiveSourceConfigKey(key string) bool {
 	if strings.Contains(value, "token") || strings.Contains(value, "secret") || strings.Contains(value, "password") {
 		return true
 	}
+	compact := strings.NewReplacer("_", "", "-", "", ".", "").Replace(value)
+	if strings.Contains(compact, "apikey") || strings.Contains(compact, "accesskey") || strings.Contains(compact, "privatekey") {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/bootstrap/app.go]: Fixed in c1ef9389a by allowing access_key_id query config while still rejecting token/secret/password/api_key/private_key/key-style secrets, with a regression test.

File: tools/linters/noinmemorydb/noinmemorydb.go [hunk 12]
```diff
@@ -52,12 +53,12 @@ func run(pass *analysis.Pass) (any, error) {
 			return
 		}
 		if sel, ok := call.Fun.(*ast.SelectorExpr); ok {
-			if (sel.Sel.Name == "Open" || sel.Sel.Name == "OpenDB") && len(call.Args) > 0 && isDatabaseSQLSelector(pass, sel) && isSQLiteDriverLiteral(call.Args[0]) {
+			if (sel.Sel.Name == "Open" || sel.Sel.Name == "OpenDB") && len(call.Args) > 0 && isDatabaseSQLSelector(pass, sel) && isSQLiteDriverLiteral(pass, call.Args[0]) {
 				report(pass, reported, call.Args[0].Pos(), call.Args[0].End())
 			}
 		}
 		for _, arg := range call.Args {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply tools/linters/noinmemorydb/noinmemorydb.go:63]: Fixed in c1ef9389a by limiting literal in-memory DSN checks to database/sql Open/OpenDB calls and adding a harmless fmt.Println(":memory:") regression case.

File: internal/bootstrap/app.go [hunk 13]
```diff
@@ -1701,52 +1705,201 @@ func sensitiveSourceConfigKey(key string) bool {
...
+		return true
+	}
 	return value == "key" || strings.HasSuffix(value, "_key")
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/bootstrap/app.go]: Fixed in c1ef9389a by no longer rejecting all *_key query fields and covering access_key_id, lookup_key, and group_key as valid non-secret config names.

File: internal/sourceruntime/service.go [hunk 14]
```diff
@@ -66,9 +70,12 @@ func (s *Service) Put(ctx context.Context, req *cerebrov1.PutSourceRuntimeReques
 		return nil, err
 	}
 	if err := source.Check(ctx, sourcecdk.NewConfig(runtime.GetConfig())); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%w: %w", ErrInvalidRequest, err)
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceruntime/service.go]: Fixed in c1ef9389a by mapping only sourcecdk.ErrInvalidConfig from runtime Check to ErrInvalidRequest; live upstream reachability errors now pass through unchanged with a regression test.

File: sdk/typescript/src/jira.ts [hunk 15]
```diff
@@ -427,10 +424,51 @@ function finding(
...
+    return normalized || undefined;
+  }
+  return undefined;
+}
+
+function objectArray(value: unknown): Array<Record<string, unknown>> {
+  if (!Array.isArray(value)) {
+    return [];
+  }
+  return value.filter((item): item is Record<string, unknown> => Boolean(item) && typeof item === "object" && !Array.isArray(item));
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sdk/typescript/src/jira.ts]: Fixed in c1ef9389a by making Jira posture arrays reject malformed entries instead of silently filtering them, with Python and TypeScript coverage.

File: internal/bootstrap/app.go [hunk 16]
```diff
@@ -1701,52 +1705,201 @@ func sensitiveSourceConfigKey(key string) bool {
...
+	case errors.Is(err, findings.ErrRuleUnavailable):
+		return connect.NewError(connect.CodeFailedPrecondition, err)
+	case errors.Is(err, findings.ErrRuntimeUnavailable):
+		return connect.NewError(connect.CodeUnavailable, err)
+	default:
+		return defaultConnectError(err)
+	}
+}
+
+func knowledgeConnectError(err error) error {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/bootstrap/app.go:1855]: Fixed in c1ef9389a by adding knowledge.ErrInvalidRequest, wrapping knowledge validation failures with it, and mapping it to InvalidArgument/400 in Connect and JSON handlers.

File: sdk/python/cerebro_sdk/jira.py [hunk 17]
```diff
@@ -408,15 +410,48 @@ def finding(
...
+def bool_value(value: Any, default: bool = False) -> str:
+    if value is None:
+        return "true" if default else "false"
+    if isinstance(value, str):
+        normalized = value.strip().lower()
+        if normalized in ("true", "1", "yes", "y", "on"):
+            return "true"
+        if normalized in ("false", "0", "no", "n", "off", ""):
+            return "false"
     return "true" if bool(value) else "false"
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sdk/python/cerebro_sdk/jira.py]: Fixed in c1ef9389a by rejecting unknown boolean strings in both Python and TypeScript Jira helpers instead of falling through to truthiness, with regression coverage.

File: cmd/cerebro/main.go [hunk 18]
```diff
@@ -296,11 +300,55 @@ func parseSourceRuntimePutArgs(args []string) (*cerebrov1.SourceRuntime, error)
...
+	if value == "" {
+		return false
+	}
+	if strings.Contains(value, "token") || strings.Contains(value, "secret") || strings.Contains(value, "password") {
+		return true
+	}
+	compact := strings.NewReplacer("_", "", "-", "", ".", "").Replace(value)
+	return compact == "key" ||
+		strings.Contains(compact, "apikey") ||
+		strings.Contains(compact, "accesskey") ||
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply cmd/cerebro/main.go]: Fixed in fff6fa1e5 by allowing literal access_key_id in CLI source config while still rejecting token/api/private key literals, with source and runtime CLI regressions.

File: internal/claims/service.go [hunk 19]
```diff
@@ -184,7 +195,19 @@ func (s *Service) ListClaims(ctx context.Context, request ListRequest) (*ListRes
 	return response, nil
 }
 
-func (s *Service) retractMissingClaims(ctx context.Context, runtimeID string, claims []*cerebrov1.Claim) (uint32, error) {
+func normalizeListLimit(limit uint32) uint32 {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/claims/service.go]: Fixed in fff6fa1e5 by preserving ListClaims limit=0 as unbounded and passing explicit large limits through unchanged, with regression coverage.

File: internal/sourceruntime/service.go [hunk 20]
```diff
@@ -43,14 +46,15 @@ func (s *Service) Put(ctx context.Context, req *cerebrov1.PutSourceRuntimeReques
...
 	}
 	if req == nil || req.GetRuntime() == nil {
-		return nil, fmt.Errorf("source runtime is required")
+		return nil, fmt.Errorf("%w: source runtime is required", ErrInvalidRequest)
 	}
 	runtime := cloneRuntime(req.GetRuntime())
 	runtime.Id = strings.TrimSpace(runtime.GetId())
 	runtime.SourceId = strings.TrimSpace(runtime.GetSourceId())
 	runtime.TenantId = strings.TrimSpace(runtime.GetTenantId())
+	clearRuntimeProgress(runtime)
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/sourceruntime/service.go]: Fixed in fff6fa1e5 by preserving caller-supplied checkpoint, next_cursor, and last_synced_at for new/restored runtimes while existing unchanged runtimes still keep stored progress.

File: sdk/typescript/src/jira.ts [hunk 21]
```diff
@@ -366,21 +366,18 @@ export function buildJiraPostureFindings(
...
     integration: "jira",
   });
-  const runtimeConfig: Record<string, string> = {};
-  const workspaceKey = options.posture.workspaceKey.trim();
-  const sourceEventId = options.posture.eventId?.trim();
-  if (workspaceKey) {
-    runtimeConfig.workspace = workspaceKey;
-  }
+  const runtimeConfig: Record<string, string> = { workspace: workspaceKey };
   await integration.ensureRuntime(runtimeConfig);
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sdk/typescript/src/jira.ts:382]: Fixed in fff6fa1e5 by building/validating Jira claims before ensureRuntime in both SDKs, so malformed posture fails before any runtime upsert side effect.

File: internal/graphingest/service.go [hunk 22]
```diff
@@ -557,6 +562,10 @@ func sensitiveConfigKey(key string) bool {
 			return true
 		}
 	}
+	compact := strings.NewReplacer("_", "", "-", "", ".", "").Replace(normalized)
+	if strings.Contains(compact, "apikey") || strings.Contains(compact, "accesskey") || strings.Contains(compact, "privatekey") {
+		return true
+	}
 	return normalized == "key" || strings.HasSuffix(normalized, "_key")
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/graphingest/service.go]: Fixed in 48f940890 by keeping non-secret selector keys like access_key_id, lookup_key, and group_key in configHash while still excluding actual secret markers, with regression coverage.

File: sdk/typescript/src/jira.ts [hunk 23]
```diff
@@ -427,10 +424,57 @@ function finding(
   };
 }
 
-function boolValue(value: boolean): string {
+function boolValue(value: unknown, defaultValue: boolean): string {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply sdk/typescript/src/jira.ts:427]: Fixed in 48f940890 by making both Python and TypeScript Jira bool helpers accept booleans/known strings only and reject object/array-style malformed values, with regression coverage.

File: internal/claims/service.go [hunk 24]
```diff
@@ -371,14 +408,20 @@ func normalizeClaim(claim *cerebrov1.Claim, runtime *cerebrov1.SourceRuntime) (*
 	case claimTypeExistence:
 	case claimTypeAttribute, claimTypeClassification:
 		if normalized.GetObjectValue() == "" {
-			return nil, fmt.Errorf("claim object value is required when claim_type=%q", normalized.GetClaimType())
+			return nil, fmt.Errorf("%w: claim object value is required when claim_type=%q", ErrInvalidRequest, normalized.GetClaimType())
 		}
 	case claimTypeRelation:
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/claims/service.go:421]: Fixed in 0aa528bc5 by validating subject URNs for every claim type before persistence/projection, with cross-tenant non-relation coverage.

File: .github/workflows/droid.yml [hunk 25]
```diff
@@ -16,51 +16,55 @@ jobs:
...
-          !contains(github.event.pull_request.body, '@droid add regression tests') &&
-          !contains(github.event.pull_request.body, '@droid scaffold source') &&
-          !contains(github.event.pull_request.title, '@droid implement') &&
-          !contains(github.event.pull_request.title, '@droid create') &&
-          !contains(github.event.pull_request.title, '@droid fix this PR') &&
-          !contains(github.event.pull_request.title, '@droid fix this pr') &&
-          !contains(github.event.pull_request.title, '@droid fix review feedback') &&
-          !contains(github.event.pull_request.title, '@droid add regression tests') &&
-          !contains(github.event.pull_request.title, '@droid scaffold source'))
+          contains(fromJSON('["OWNER","MEMBER","COLLABORATOR"]'), github.event.pull_request.author_association) &&
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply .github/workflows/droid.yml]: Fixed in 0aa528bc5 by removing the PR-author association gate from pull_request title/body triggers; same-repo and trusted-actor checks still gate execution on the editing actor.

File: internal/claims/service.go [hunk 26]
```diff
@@ -371,14 +408,20 @@ func normalizeClaim(claim *cerebrov1.Claim, runtime *cerebrov1.SourceRuntime) (*
 	case claimTypeExistence:
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/claims/service.go:416]: Fixed in 0aa528bc5 by validating object URNs whenever present, including object_ref on non-relation claims, preventing cross-tenant projected entity upserts.

File: sources/okta/source.go [hunk 27]
```diff
@@ -702,7 +830,9 @@ func (s *Source) getJSON(ctx context.Context, settings settings, requestPath str
 
 	client := s.client
 	if client == nil {
-		client = http.DefaultClient
+		client = oktaHTTPClientNoRedirect(nil)
```
Comment (jonathanhaaswriter) [inline sources/okta/source.go]: [P1] Add resolved-host SSRF checks to Okta HTTP requests

Okta `normalizeDomain` / `normalizeBaseURL` reject literal private/loopback hosts, but `getJSON` sends requests with `client.Do` using a plain no-redirect client and never re-checks the resolved IP. A hostname like `okta.attacker.example` can pass validation and later resolve to `169.254.169.254` or another private address, causing requests with the `SSWS` token to be sent to internal targets. Wrap Okta’s transport with the same resolved-host guard used by GitHub, or resolve and reject unsafe IPs before the request.

Comment (jonathanhaaswriter) [reply sources/okta/source.go]: Fixed in 40e02cafb by wrapping Okta HTTP transport with resolved-host validation that rejects private/link-local resolutions and fails closed on DNS errors before sending SSWS-authenticated requests. Added regression tests for both unsafe resolved IPs and resolution failures.

File: cmd/cerebro/main.go [hunk 28]
```diff
@@ -296,11 +300,54 @@ func parseSourceRuntimePutArgs(args []string) (*cerebrov1.SourceRuntime, error)
...
+		if err != nil {
+			return nil, err
+		}
+		runtime.Config[key] = resolved
 	}
 	return runtime, nil
 }
 
+func sourceConfigValueFromArg(key string, value string) (string, error) {
+	if strings.HasPrefix(value, "env:") {
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply cmd/cerebro/main.go:313]: Fixed in 9103e3f17 by treating env: as an environment reference only for sensitive CLI config keys; non-sensitive fields such as phrase/filter/search preserve env:-prefixed literals. Added CLI regression coverage.
